File: regress/tests/6_fixed_bugs/UsedPointerSweeped/UsedPointerSweeped.c
#include <stdio.h>
#include <stdlib.h>
// return costant value
int get_int(void) {  return 42; }

// return constant value
const char* get_str(void) {
  return "So Long, and Thanks for All the Fish";
}

void* make_entry(void) {
  /* allocate some pointer */
  return (void*)0xdeadbeaf;
}

void dump(void* p) {
  if(p != 0xdeadbeaf) {
    printf("wow! someone overwrite my pointer!(%x)\n", p);
    abort();
  }
}

File: regress/tests/6_fixed_bugs/UsedPointerSweeped/UsedPointerSweeped.hs
module Main where

import Foreign.Ptr
import Foreign.C.String
import Control.Monad

foreign import primitive "const.sizeof(int)" msgSize :: Int

-- return constant value to disable optimaization
foreign import ccall "get_int" getInt :: IO Int
foreign import ccall "get_str" getStr :: IO CString

-- allocate pointer
foreign import ccall "make_entry" make :: IO (Ptr Word8)
-- dump the value
foreign import ccall "dump" dump :: Ptr a -> IO ()

loop _ = do entry <- make -- allocate a pointer
            forever $ thread entry

thread msg = do n <- getInt
                if n < msgSize then
                  return ()
                else
                  do dump msg  -- dump a allocated pointer
                     str <- getStr
                     token <- peekCString str
                     thread msg -- and repeat this

main = loop ()
